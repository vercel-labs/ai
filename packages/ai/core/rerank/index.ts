export * from './rerank';
export * from './rerank-result';
