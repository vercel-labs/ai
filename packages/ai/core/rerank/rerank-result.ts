import { RerankedDocumentIndex } from '../types';
import { RerankingModelUsage } from '../types/usage';

/**
The result of an `rerank` call.
It contains the documents, the reranked indices, and additional information.
 */
export interface RerankResult<VALUE> {
  /**
  The documents that were reranked.
     */
  readonly documents: Array<VALUE>;

  /**
  The reranked indices.
    */
  readonly rerankedIndices: Array<RerankedDocumentIndex>;

  /**
   * The reranked documents.
   * Only available if `returnDocuments` was set to `true`.
   * The order of the documents is the same as the order of the indices.
   */
  readonly rerankedDocuments?: Array<VALUE>;

  /**
  The reranking token usage.
    */
  readonly usage: RerankingModelUsage;

  /**
  Optional raw response data.
     */
  readonly rawResponse?: {
    /**
  Response headers.
       */
    headers?: Record<string, string>;
  };
}
