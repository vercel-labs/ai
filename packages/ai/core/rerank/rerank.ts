import { RerankingModelV1DocumentIndex } from '@ai-sdk/provider';
import { retryWithExponentialBackoff } from '../../util/retry-with-exponential-backoff';
import { assembleOperationName } from '../telemetry/assemble-operation-name';
import { getBaseTelemetryAttributes } from '../telemetry/get-base-telemetry-attributes';
import { getTracer } from '../telemetry/get-tracer';
import { recordSpan } from '../telemetry/record-span';
import { selectTelemetryAttributes } from '../telemetry/select-telemetry-attributes';
import { TelemetrySettings } from '../telemetry/telemetry-settings';
import { RerankingModel } from '../types';
import { splitArray } from '../util/split-array';
import { RerankResult } from './rerank-result';

/**
Rerank documents using an reranking model. The type of the value is defined by the reranking model.

@param model - The Reranking model to use.
@param values - The documents that should be reranking.
@param query - The query is a string that represents the query to rerank the documents against.
@param topK - Top k documents to rerank.
@param returnDocuments - Return the reranked documents in the response (In same order as indices).

@param maxRetries - Maximum number of retries. Set to 0 to disable retries. Default: 2.
@param abortSignal - An optional abort signal that can be used to cancel the call.
@param headers - Additional HTTP headers to be sent with the request. Only applicable for HTTP-based providers.

@returns A result object that contains the reranked documents, the reranked indices, and additional information.
 */
export async function rerank<VALUE>({
  model,
  values,
  query,
  topK,
  returnDocuments = false,
  maxRetries,
  abortSignal,
  headers,
  experimental_telemetry: telemetry,
}: {
  /**
The reranking model to use.
  */
  model: RerankingModel<VALUE>;

  /**
The documents that should be reranked.
   */
  values: Array<VALUE>;

  /**
The query is a string that represents the query to rerank the documents against.
   */
  query: string;

  /**
Top k documents to rerank.
  */
  topK: number;

  /**
Return the reranked documents in the response (In same order as indices).

@default false
   */
  returnDocuments?: boolean;

  /**
Maximum number of retries per reranking model call. Set to 0 to disable retries.

@default 2
   */
  maxRetries?: number;

  /**
Abort signal.
 */
  abortSignal?: AbortSignal;

  /**
Additional headers to include in the request.
Only applicable for HTTP-based providers.
 */
  headers?: Record<string, string>;

  /**
   * Optional telemetry configuration (experimental).
   */
  experimental_telemetry?: TelemetrySettings;
}): Promise<RerankResult<VALUE>> {
  const baseTelemetryAttributes = getBaseTelemetryAttributes({
    model,
    telemetry,
    headers,
    settings: { maxRetries },
  });

  const tracer = getTracer(telemetry);

  return recordSpan({
    name: 'ai.rerank',
    attributes: selectTelemetryAttributes({
      telemetry,
      attributes: {
        ...assembleOperationName({ operationId: 'ai.rerank', telemetry }),
        ...baseTelemetryAttributes,
        'ai.values': { input: () => JSON.stringify(values) },
      },
    }),
    tracer,
    fn: async span => {
      const retry = retryWithExponentialBackoff({ maxRetries });
      const maxDocumentsPerCall = model.maxDocumentsPerCall;

      if (maxDocumentsPerCall == null) {
        const { rerankedIndices, usage, rerankedDocuments, rawResponse } =
          await retry(() =>
            recordSpan({
              name: 'ai.rerank.doRerank',
              attributes: selectTelemetryAttributes({
                telemetry,
                attributes: {
                  ...assembleOperationName({
                    operationId: 'ai.rerank.doRerank',
                    telemetry,
                  }),
                  ...baseTelemetryAttributes,
                  // specific settings that only make sense on the outer level:
                  'ai.values': { input: () => [JSON.stringify(values)] },
                },
              }),
              tracer,
              fn: async doRerankSpan => {
                const modelResponse = await model.doRerank({
                  values,
                  query,
                  topK,
                  returnDocuments,
                  abortSignal,
                  headers,
                });

                const rerankedIndices = modelResponse.rerankedIndices;
                const usage = modelResponse.usage ?? { tokens: NaN };
                const rerankedDocuments = modelResponse.rerankedDocuments ?? [];

                doRerankSpan.setAttributes(
                  selectTelemetryAttributes({
                    telemetry,
                    attributes: {
                      'ai.rerankedIndices': {
                        output: () => JSON.stringify(rerankedIndices),
                      },
                      'ai.usage.tokens': usage.tokens,
                      'ai.rerankedDocuments': {
                        output: () => JSON.stringify(rerankedDocuments),
                      },
                    },
                  }),
                );

                return {
                  rerankedIndices,
                  usage,
                  rerankedDocuments,
                  rawResponse: modelResponse.rawResponse,
                };
              },
            }),
          );

        span.setAttributes(
          selectTelemetryAttributes({
            telemetry,
            attributes: {
              'ai.rerankedIndices': {
                output: () => JSON.stringify(rerankedIndices),
              },
              'ai.rerankedDocuments': {
                output: () => JSON.stringify(rerankedDocuments),
              },
              'ai.usage.tokens': usage.tokens,
            },
          }),
        );

        return new DefaultRerankResult({
          values,
          rerankedIndices,
          rerankedDocuments,
          usage,
          rawResponse,
        });
      }

      // split the values into chunks that are small enough for the model:
      const valueChunks = splitArray(values, maxDocumentsPerCall);

      const rerankedIndices: Array<RerankingModelV1DocumentIndex> = [];
      const rerankedDocuments: Array<VALUE> = [];
      let tokens = 0;

      for (const chunk of valueChunks) {
        const {
          rerankedIndices: chunkIndices,
          rerankedDocuments: chunkedRerankedDocuments,
          usage,
        } = await retry(() =>
          recordSpan({
            name: 'ai.rerank.doRerank',
            attributes: selectTelemetryAttributes({
              telemetry,
              attributes: {
                ...assembleOperationName({
                  operationId: 'ai.rerank.doRerank',
                  telemetry,
                }),
                ...baseTelemetryAttributes,
                'ai.values': { input: () => [JSON.stringify(chunk)] },
              },
            }),
            tracer,
            fn: async doRerankSpan => {
              const modelResponse = await model.doRerank({
                values: chunk,
                query,
                topK,
                returnDocuments,
                abortSignal,
                headers,
              });

              const chunkIndices = modelResponse.rerankedIndices;
              const usage = modelResponse.usage ?? { tokens: NaN };
              const chunkedRerankedDocuments =
                modelResponse.rerankedDocuments ?? [];

              doRerankSpan.setAttributes(
                selectTelemetryAttributes({
                  telemetry,
                  attributes: {
                    'ai.rerankedIndices': {
                      output: () => JSON.stringify(chunkIndices),
                    },
                    'ai.usage.tokens': usage.tokens,
                    'ai.rerankedDocuments': {
                      output: () => JSON.stringify(chunkedRerankedDocuments),
                    },
                  },
                }),
              );

              return {
                rerankedIndices: chunkIndices,
                usage,
                rerankedDocuments: chunkedRerankedDocuments,
              };
            },
          }),
        );

        rerankedIndices.push(...chunkIndices);
        rerankedDocuments.push(...chunkedRerankedDocuments);
        tokens += usage.tokens;
      }

      span.setAttributes(
        selectTelemetryAttributes({
          telemetry,
          attributes: {
            'ai.rerankedIndices': {
              output: () => JSON.stringify(rerankedIndices),
            },
            'ai.rerankedDocuments': {
              output: () => JSON.stringify(rerankedDocuments),
            },
            'ai.usage.tokens': tokens,
          },
        }),
      );

      return new DefaultRerankResult({
        values,
        rerankedIndices,
        rerankedDocuments,
        usage: { tokens },
      });
    },
  });
}

class DefaultRerankResult<VALUE> implements RerankResult<VALUE> {
  readonly documents: RerankResult<VALUE>['documents'];
  readonly rerankedIndices: RerankResult<VALUE>['rerankedIndices'];
  readonly rerankedDocuments: RerankResult<VALUE>['documents'];
  readonly usage: RerankResult<VALUE>['usage'];
  readonly rawResponse: RerankResult<VALUE>['rawResponse'];

  constructor(options: {
    values: RerankResult<VALUE>['documents'];
    rerankedIndices: RerankResult<VALUE>['rerankedIndices'];
    rerankedDocuments: RerankResult<VALUE>['documents'];
    usage: RerankResult<VALUE>['usage'];
    rawResponse?: RerankResult<VALUE>['rawResponse'];
  }) {
    this.documents = options.values;
    this.rerankedIndices = options.rerankedIndices;
    this.rerankedDocuments = options.rerankedDocuments;
    this.usage = options.usage;
    this.rawResponse = options.rawResponse;
  }
}
