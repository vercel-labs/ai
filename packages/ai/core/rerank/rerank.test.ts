import assert from 'node:assert';
import { MockTracer } from '../test/mock-tracer';
import { rerank } from './rerank';
import {
  mockRerank,
  MockRerankingModelV1,
} from '../test/mock-reranking-model-v1';

const dummyDocumentsIndices = [1, 0];
const dummyDocuments = ['sunny day at the beach', 'rainy day in the city'];
const dummyRerankedDocuments = [
  'rainy day in the city',
  'sunny day at the beach',
];
const query = 'rainy day';
const topK = 2;

describe('result.reranking', () => {
  it('should reranking documents', async () => {
    const result = await rerank({
      model: new MockRerankingModelV1({
        maxDocumentsPerCall: 5,
        doRerank: mockRerank(
          dummyDocuments,
          dummyDocumentsIndices,
          dummyRerankedDocuments,
        ),
      }),
      values: dummyDocuments,
      query,
      topK,
    });

    assert.deepStrictEqual(result.rerankedIndices, dummyDocumentsIndices);
  });

  it('should reranking documents when several calls are required', async () => {
    let callCount = 0;
    const result = await rerank({
      model: new MockRerankingModelV1({
        maxDocumentsPerCall: 1,
        doRerank: async () => {
          switch (callCount++) {
            case 0:
              return {
                rerankedIndices: dummyDocumentsIndices.slice(0, 1),
              };
            case 1:
              return {
                rerankedIndices: dummyDocumentsIndices.slice(1),
              };
            default:
              throw new Error('Unexpected call');
          }
        },
      }),
      values: dummyDocuments,
      query,
      topK,
    });

    assert.deepStrictEqual(result.rerankedIndices, dummyDocumentsIndices);
  });
});

describe('result.value', () => {
  it('should include value in the result', async () => {
    const result = await rerank({
      model: new MockRerankingModelV1({
        maxDocumentsPerCall: 5,
        doRerank: mockRerank(
          dummyDocuments,
          dummyDocumentsIndices,
          dummyRerankedDocuments,
        ),
      }),
      values: dummyDocuments,
      query,
      topK,
      returnDocuments: true,
    });

    assert.deepStrictEqual(result.rerankedDocuments, dummyRerankedDocuments);
  });
});

describe('result.usage', () => {
  it('should include usage in the result', async () => {
    let callCount = 0;
    const result = await rerank({
      model: new MockRerankingModelV1({
        maxDocumentsPerCall: 1,
        doRerank: async () => {
          switch (callCount++) {
            case 0:
              return {
                rerankedIndices: dummyDocumentsIndices.slice(0, 1),
                usage: { tokens: 10 },
              };
            case 1:
              return {
                rerankedIndices: dummyDocumentsIndices.slice(1),
                usage: { tokens: 10 },
              };
            default:
              throw new Error('Unexpected call');
          }
        },
      }),
      values: dummyDocuments,
      query,
      topK,
    });

    assert.deepStrictEqual(result.usage, { tokens: 20 });
  });
});

describe('options.headers', () => {
  it('should set headers', async () => {
    const result = await rerank({
      model: new MockRerankingModelV1({
        maxDocumentsPerCall: 5,
        doRerank: async ({ headers }) => {
          assert.deepStrictEqual(headers, {
            'custom-request-header': 'request-header-value',
          });

          return { rerankedIndices: dummyDocumentsIndices };
        },
      }),
      values: dummyDocuments,
      query,
      topK,
      headers: { 'custom-request-header': 'request-header-value' },
    });

    assert.deepStrictEqual(result.rerankedIndices, dummyDocumentsIndices);
  });
});

describe('telemetry', () => {
  let tracer: MockTracer;

  beforeEach(() => {
    tracer = new MockTracer();
  });

  it('should not record any telemetry data when not explicitly enabled', async () => {
    await rerank({
      model: new MockRerankingModelV1({
        maxDocumentsPerCall: 5,
        doRerank: mockRerank(
          dummyDocuments,
          dummyDocumentsIndices,
          dummyRerankedDocuments,
        ),
      }),
      values: dummyDocuments,
      query,
      topK,
    });
    assert.deepStrictEqual(tracer.jsonSpans, []);
  });

  it('should record telemetry data when enabled (single call path)', async () => {
    await rerank({
      model: new MockRerankingModelV1({
        maxDocumentsPerCall: null,
        doRerank: mockRerank(
          dummyDocuments,
          dummyDocumentsIndices,
          dummyRerankedDocuments,
          {
            tokens: 10,
          },
        ),
      }),
      values: dummyDocuments,
      query,
      topK,
      experimental_telemetry: {
        isEnabled: true,
        functionId: 'test-function-id',
        metadata: {
          test1: 'value1',
          test2: false,
        },
        tracer,
      },
    });

    expect(tracer.jsonSpans).toMatchSnapshot();
  });

  it('should record telemetry data when enabled (multiple calls path)', async () => {
    let callCount = 0;
    await rerank({
      model: new MockRerankingModelV1({
        maxDocumentsPerCall: 1,
        doRerank: async ({ values }) => {
          switch (callCount++) {
            case 0:
              assert.deepStrictEqual(values, dummyDocuments.slice(0, 1));
              return {
                rerankedIndices: dummyDocumentsIndices.slice(0, 1),
                usage: { tokens: 10 },
              };
            case 1:
              assert.deepStrictEqual(values, dummyDocuments.slice(1));
              return {
                rerankedIndices: dummyDocumentsIndices.slice(1),
                usage: { tokens: 10 },
              };
            default:
              throw new Error('Unexpected call');
          }
        },
      }),
      values: dummyDocuments,
      query,
      topK,
      experimental_telemetry: {
        isEnabled: true,
        functionId: 'test-function-id',
        metadata: {
          test1: 'value1',
          test2: false,
        },
        tracer,
      },
    });

    expect(tracer.jsonSpans).toMatchSnapshot();
  });

  it('should not record telemetry inputs / outputs when disabled', async () => {
    await rerank({
      model: new MockRerankingModelV1({
        maxDocumentsPerCall: null,
        doRerank: mockRerank(
          dummyDocuments,
          dummyDocumentsIndices,
          dummyRerankedDocuments,
          {
            tokens: 10,
          },
        ),
      }),
      values: dummyDocuments,
      query,
      topK,
      experimental_telemetry: {
        isEnabled: true,
        recordInputs: false,
        recordOutputs: false,
        tracer,
      },
    });

    expect(tracer.jsonSpans).toMatchSnapshot();
  });
});
