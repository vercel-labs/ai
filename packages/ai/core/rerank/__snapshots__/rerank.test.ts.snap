// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`telemetry > should not record telemetry inputs / outputs when disabled 1`] = `
[
  {
    "attributes": {
      "ai.model.id": "mock-model-id",
      "ai.model.provider": "mock-provider",
      "ai.operationId": "ai.rerank",
      "ai.usage.tokens": 10,
      "operation.name": "ai.rerank",
    },
    "events": [],
    "name": "ai.rerank",
  },
  {
    "attributes": {
      "ai.model.id": "mock-model-id",
      "ai.model.provider": "mock-provider",
      "ai.operationId": "ai.rerank.doRerank",
      "ai.usage.tokens": 10,
      "operation.name": "ai.rerank.doRerank",
    },
    "events": [],
    "name": "ai.rerank.doRerank",
  },
]
`;

exports[`telemetry > should record telemetry data when enabled (multiple calls path) 1`] = `
[
  {
    "attributes": {
      "ai.model.id": "mock-model-id",
      "ai.model.provider": "mock-provider",
      "ai.operationId": "ai.rerank",
      "ai.rerankedDocuments": "[]",
      "ai.rerankedIndices": "[1,0]",
      "ai.telemetry.functionId": "test-function-id",
      "ai.telemetry.metadata.test1": "value1",
      "ai.telemetry.metadata.test2": false,
      "ai.usage.tokens": 20,
      "ai.values": "["sunny day at the beach","rainy day in the city"]",
      "operation.name": "ai.rerank test-function-id",
      "resource.name": "test-function-id",
    },
    "events": [],
    "name": "ai.rerank",
  },
  {
    "attributes": {
      "ai.model.id": "mock-model-id",
      "ai.model.provider": "mock-provider",
      "ai.operationId": "ai.rerank.doRerank",
      "ai.rerankedDocuments": "[]",
      "ai.rerankedIndices": "[1]",
      "ai.telemetry.functionId": "test-function-id",
      "ai.telemetry.metadata.test1": "value1",
      "ai.telemetry.metadata.test2": false,
      "ai.usage.tokens": 10,
      "ai.values": [
        "["sunny day at the beach"]",
      ],
      "operation.name": "ai.rerank.doRerank test-function-id",
      "resource.name": "test-function-id",
    },
    "events": [],
    "name": "ai.rerank.doRerank",
  },
  {
    "attributes": {
      "ai.model.id": "mock-model-id",
      "ai.model.provider": "mock-provider",
      "ai.operationId": "ai.rerank.doRerank",
      "ai.rerankedDocuments": "[]",
      "ai.rerankedIndices": "[0]",
      "ai.telemetry.functionId": "test-function-id",
      "ai.telemetry.metadata.test1": "value1",
      "ai.telemetry.metadata.test2": false,
      "ai.usage.tokens": 10,
      "ai.values": [
        "["rainy day in the city"]",
      ],
      "operation.name": "ai.rerank.doRerank test-function-id",
      "resource.name": "test-function-id",
    },
    "events": [],
    "name": "ai.rerank.doRerank",
  },
]
`;

exports[`telemetry > should record telemetry data when enabled (single call path) 1`] = `
[
  {
    "attributes": {
      "ai.model.id": "mock-model-id",
      "ai.model.provider": "mock-provider",
      "ai.operationId": "ai.rerank",
      "ai.rerankedDocuments": "["rainy day in the city","sunny day at the beach"]",
      "ai.rerankedIndices": "[1,0]",
      "ai.telemetry.functionId": "test-function-id",
      "ai.telemetry.metadata.test1": "value1",
      "ai.telemetry.metadata.test2": false,
      "ai.usage.tokens": 10,
      "ai.values": "["sunny day at the beach","rainy day in the city"]",
      "operation.name": "ai.rerank test-function-id",
      "resource.name": "test-function-id",
    },
    "events": [],
    "name": "ai.rerank",
  },
  {
    "attributes": {
      "ai.model.id": "mock-model-id",
      "ai.model.provider": "mock-provider",
      "ai.operationId": "ai.rerank.doRerank",
      "ai.rerankedDocuments": "["rainy day in the city","sunny day at the beach"]",
      "ai.rerankedIndices": "[1,0]",
      "ai.telemetry.functionId": "test-function-id",
      "ai.telemetry.metadata.test1": "value1",
      "ai.telemetry.metadata.test2": false,
      "ai.usage.tokens": 10,
      "ai.values": [
        "["sunny day at the beach","rainy day in the city"]",
      ],
      "operation.name": "ai.rerank.doRerank test-function-id",
      "resource.name": "test-function-id",
    },
    "events": [],
    "name": "ai.rerank.doRerank",
  },
]
`;
